source/agora/utils/SCPPrettyPrinter.d

```diff
@@ -0,0 +1,606 @@
+    {
+        if (auto qset = hash in qmap)
+            return *qset;
+
+        return SCPQuorumSetPtr.init;
+    }
+
+    SCPEnvelope env;
+    env.statement.nodeID = NodeID(StellarHash(pair.address[]));
+
+    /** SCP PREPARE */
+    env.statement.pledges.type_ = SCPStatementType.SCP_ST_PREPARE;
+    env.statement.pledges.prepare_ = SCPStatement._pledges_t._prepare_t.init; // must initialize
+    env.statement.pledges.prepare_.quorumSetHash = quorum_hash;
+    env.statement.pledges.prepare_.ballot = ballot;
+    env.statement.pledges.prepare_.nC = 100;
+    env.statement.pledges.prepare_.nH = 200;
+
+    // missing signature
+    env.signature = typeof(env.signature).init;
+
+    // missing signature
+    static immutable MissingSig = `{ statement: { node: GBUV...KOEK, slotIndex: 0, pledge: SCP_PREPARE { qset: { hash: 0xfee13ea2fd956028d31b93cbf0a47ea97c6c9868b7e911b797667113ed12bf0d, quorum: <unknown> }, ballot: { counter: 42, value: { tx_set: [Type : Payment, Inputs (1): 0x0000...0000[0]:0x0000...0000
+Outputs (8): GCOQ...LRIJ(62,500,000), GCOQ...LRIJ(62,500,000), GCOQ...LRIJ(62,500,000),
+GCOQ...LRIJ(62,500,000), GCOQ...LRIJ(62,500,000), GCOQ...LRIJ(62,500,000),
+GCOQ...LRIJ(62,500,000), GCOQ...LRIJ(62,500,000)], enrolls: [{ utxo: 0x0000...e26f, seed: 0x4a5e...a33b, cycles: 1008, sig: 0x0000...be78 }, { utxo: 0x0000...e26f, seed: 0x4a5e...a33b, cycles: 1008, sig: 0x0000...be78 }] } }, prep: <null>, prepPrime: <null>, nc: 100, nH: 200 } }, sig: 0x0000...0000 }`;
+
+    assert(MissingSig == format("%s", scpPrettify(env)),
+                         format("%s", scpPrettify(env)));
+
+    env.signature = pair.secret.sign(hashFull(0)[])[].toVec();
+
+    // null quorum (hash not found)
+    static immutable PrepareRes1 = `{ statement: { node: GBUV...KOEK, slotIndex: 0, pledge: SCP_PREPARE { qset: { hash: 0xfee13ea2fd956028d31b93cbf0a47ea97c6c9868b7e911b797667113ed12bf0d, quorum: <unknown> }, ballot: { counter: 42, value: { tx_set: [Type : Payment, Inputs (1): 0x0000...0000[0]:0x0000...0000
+Outputs (8): GCOQ...LRIJ(62,500,000), GCOQ...LRIJ(62,500,000), GCOQ...LRIJ(62,500,000),
+GCOQ...LRIJ(62,500,000), GCOQ...LRIJ(62,500,000), GCOQ...LRIJ(62,500,000),
+GCOQ...LRIJ(62,500,000), GCOQ...LRIJ(62,500,000)], enrolls: [{ utxo: 0x0000...e26f, seed: 0x4a5e...a33b, cycles: 1008, sig: 0x0000...be78 }, { utxo: 0x0000...e26f, seed: 0x4a5e...a33b, cycles: 1008, sig: 0x0000...be78 }] } }, prep: <null>, prepPrime: <null>, nc: 100, nH: 200 } }, sig: 0x0af7...b5ab }`;
+
+    // with quorum mapping
+    static immutable PrepareRes2 = `{ statement: { node: GBUV...KOEK, slotIndex: 0, pledge: SCP_PREPARE { qset: { hash: 0xfee13ea2fd956028d31b93cbf0a47ea97c6c9868b7e911b797667113ed12bf0d, quorum: QuorumConfig(2, [GBFDLGQQDDE2CAYVELVPXUXR572ZT5EOTMGJQBPTIHSLPEOEZYQQCEWN, GBYK4I37MZKLL4A2QS7VJCTDIIJK7UXWQWKXKTQ5WZGT2FPCGIVIQCY5], []) }, ballot: { counter: 42, value: { tx_set: [Type : Payment, Inputs (1): 0x0000...0000[0]:0x0000...0000
```

Comment:
Should we abbreviate `SCP_PREPARE.qset.hash` ? Also `QuorumConfig` entries ?

Reply:
Good ideas.